src/include/daos/cont_props.h

```diff
@@ -158,4 +158,10 @@ daos_cont_prop2global_version(daos_prop_t *prop);
 uint32_t
 daos_cont_prop2obj_version(daos_prop_t *prop);
 
+static inline uint32_t
+daos_cont_props2pda(struct cont_props *props, bool is_ec_obj)
+{
+	return is_ec_obj ? props->dcp_ec_pda : props->dcp_rp_pda;
```

Comment:
Just curious, what is the difference between ec_pda and rp_pda?

Reply:
it is just for EC object and replica obj. -
141 /**                                                                                                 
142  * the placement algorithm should place two-way and three-way                                       
143  * replication object within a PD; for those object classes with                                    
144  * more than 3 replicas, DAOS will place three replicas within a PD                                 
145  * and switch to another PD.                                                                        
146  */                                                                                                 
147 #define DAOS_PROP_PO_RP_PDA_DEFAULT>    3                                                           
148 /**                                                                                                 
149  * the placement algorithm always tries to scatter shards of EC                                     
150  * object to different PDs.                                                                         
151  */                                                                                                 
152 #define DAOS_PROP_PO_EC_PDA_DEFAULT>    1

Reply:
>the placement algorithm should place two-way and three-way
143 * replication object within a PD;

So this will favor write? not read?

>the placement algorithm always tries to scatter shards of EC
150 * object to different PDs.

But this will favor read? not write?  Or I miss sth?

Reply:
> > the placement algorithm should place two-way and three-way
> > 143 * replication object within a PD;
> 
> So this will favor write? not read?

Write could be faster as IO forward is faster, for read it is not bad right? as any replica can be read.
Rebuild could be better also as fetch within PD.

> 
> > the placement algorithm always tries to scatter shards of EC
> > 150 * object to different PDs.
> 
> But this will favor read? not write? Or I miss sth?

EC possibly need to read from all shards, so spread to multiple PD?
anyway it is not a problem of this patch.

Reply:
>Write could be faster as IO forward is faster, for read it is not bad right? as any replica can be read.
Rebuild could be better also as fetch within PD.

Then any replica read will all comes from 1 PD, i.e. using bandwidth from single PD.

> EC possibly need to read from all shards, so spread to multiple PD?

So EC  multiples shards will spread to multiple PD for read? so why not for replica objects?  Anyway I do not see how it is different between EC and replica?

Reply:
> > Write could be faster as IO forward is faster, for read it is not bad right? as any replica can be read.
> > Rebuild could be better also as fetch within PD.
> 
> Then any replica read will all comes from 1 PD, i.e. using bandwidth from single PD.
> 
> > EC possibly need to read from all shards, so spread to multiple PD?
> 
> So EC multiples shards will spread to multiple PD for read? so why not for replica objects? Anyway I do not see how it is different between EC and replica?

below is quoted from liang's HLD -
Besides adaptive routing during regular I/O, another thing that should be considered for object placement is switch-to-switch hops of server-side message forwarding for two-phase commit write, and data transferring for rebuilding objects.
Distributed I/O of DAOS is primary-slave model, where the client only sends RPC to the primary server of a redundancy group, and then the primary server forwards the RPC and RDMA descriptor to other servers. For a large EC object, avoiding one or few objects consume all bandwidth of a performance domain is the main concern, so spreading shards to many groups can avoid bandwidth hogging and benefit overall performance. However, for an object with a small redundancy group size (e.g., three-way replicated object), bandwidth consumption is already limited by shards (targets) number. Instead of spreading shards to different performance domains, placing shards of this type of object in the same performance domain is the optimal option because it has a few benefits:
• Avoid extra network-hops of forwarding RPCs between servers and reduce latency of regular I/O.
• Reduce data transfer between DAOS groups during rebuild.

Reply:
> > Write could be faster as IO forward is faster, for read it is not bad right? as any replica can be read.
> > Rebuild could be better also as fetch within PD.
> 
> Then any replica read will all comes from 1 PD, i.e. using bandwidth from single PD.
> 
> > EC possibly need to read from all shards, so spread to multiple PD?
> 
> So EC multiples shards will spread to multiple PD for read? so why not for replica objects? Anyway I do not see how it is different between EC and replica?

Replication is mostly for small objects, so it's latency sensitive, whereas EC is for large object which want more bandwidth

Reply:
Well, multiple groups(GX, SX) also intends for scalability and bandwidth, which can have smaller replication number.  So we should also choose multiple PD for this case as well?

@liuxuezhao  how do we choose PD for this case?

Reply:
Indeed, though EC aggregation and rebuild still need cross the PD.